### DISCUSSION TITLE
[Question] External library support?

### DISCUSSION THREAD

**Discussion by R-Bower** [vanilla-extract-css/vanilla-extract#717] [Q&A]
I've read through the docs and examples but haven't seen anything related to external library support.  Here's an example of what I mean:

For the sake of brevity, let's assume that I've setup a `sprinkles.css.ts` in an external package with the following properties:

```
// file defined in @my-lib/styles
const responsiveProperties = defineProperties({
  conditions: {
    mobile: {},
    tablet: { '@media': 'screen and (min-width: 768px)' },
    desktop: { '@media': 'screen and (min-width: 1024px)' },
  },
  defaultCondition: 'mobile',
  properties: {
    display: ['none', 'block', 'flex'],
    flexDirection: ['row', 'column'],
  }
});

export const flexSprinkles = createSprinkles(
  responsiveProperties,
);
```

I have an external application that depends on the above `@my-lib/styles` package.

```
// @my-app/frontend
import {flexSprinkles} from "@my-lib/styles"
// ...
<div className={flexSprinkles({
  display: "flex",
  flexDirection: "column",
})}></div>
```

Is there a way to make this work as expected?  Ideally, I'd like to define reusable "sprinkles" exports for use in separate projects.  I'm aware of packages like https://github.com/TheMightyPenguin/dessert-box, but these seem to be react-specific.  Ideally, this solution should work in non-react libraries as well.  

Nice work btw. I love what I've seen so far.

**Comment by mattcompiles** [comment]
You can definitely do this, you just have to decide where you want to do the compilation. You can either make it a requirement of your consumers to run the vanilla-extract plugin of their choice. Or you can pre-compile at the library level to CSS files and expect that your consumers can compile `.css` files as part of their application build. If you choose the latter than the `rollup-plugin` is the easiest existing option to achieve this.

**Reply by riccardoperra** [reply]
Hi @mattcompiles, I have some doubts too.

I also have a monorepo setup with an app (vite) and ui package (rollup), but I couldn't manage to get it working.  What I'm trying to achieve is to define a base theme contract with `createThemeContract` in the library package, and consume in my application defining a new theme based of it.

An example of my use case in a larger codebase using Vite for both consumer application and external package:
[theme.css.ts - external package](https://github.com/riccardoperra/codeimage/blob/8c5135861293c37d7236c5aa73e3d89eab1b877b/packages/ui/src/lib/theme/theme.css.ts#L3)
[theme1 - consumer app](https://github.com/riccardoperra/codeimage/blob/main/apps/codeimage/src/theme/dark-theme.css.ts)


Here's a small example https://github.com/riccardoperra/solid-vanilla-extract-repro where I am trying to testing everything with rollup

## Bundling only as part of application build
- Consumer app vite configuration: [vite.config.ts](https://github.com/riccardoperra/solid-vanilla-extract-repro/blob/rollup_plugin_no_ve/packages/example-1/vite.config.ts)
- Library configuration: [rollup.config.js](https://github.com/riccardoperra/solid-vanilla-extract-repro/blob/rollup_plugin_no_ve/packages/ui-rollup/rollup.config-no-ve.js)

If I do that, it seems that during css build phase vanilla extract tries to resolve solid-js imports for "node", but I don't have ssr enabled. This forced me to patch their imports momentarily in node_modules by replacing them with the browser ones. This is obviously a workaround because doing so cannot use ssr.

<details>
<summary>Screenshot error</summary>

![image](https://user-images.githubusercontent.com/37072694/171809981-bbb8c9db-03c3-48f9-90ca-2ef7fe627cbf.png)

</details>

## Bundling both application and ui library
- Consumer app vite configuration: [vite.config.ts](https://github.com/riccardoperra/solid-vanilla-extract-repro/blob/rollup_plugin_no_ve/packages/example-1/vite.config.ts)
- Library configuration: [rollup.config.js](https://github.com/riccardoperra/solid-vanilla-extract-repro/blob/rollup_plugin_no_ve/packages/ui-rollup/rollup.config.js)

Same as before, everything seems working fine if I consume components, but if I try to import themes or styles from the library package this is the error that I have

<details>
<summary>Screenshot error</summary>

![image](https://user-images.githubusercontent.com/37072694/171811508-987f9b95-490b-493f-a657-fb8867092c84.png)

</details>

---

Having said that, I don't understand if I am looking for something that is impossible to do at the moment or if I am wrong configurations / approach.

> **Note** Using vite in the consumer application without bundling the css inside the library, I noticed that even using your branch "vite-resolution-fix" I have the same 404 error, I had to move up to @vanilla-extract/vite-plugin@3.1.4
<details>
<summary>Screenshot error</summary>

![image](https://user-images.githubusercontent.com/37072694/171813447-dbde1ed1-fc76-43ef-ac52-2b2feae7403e.png)

</details>


https://github.com/riccardoperra/solid-vanilla-extract-repro/blob/ac21f9b6b9c6664759aa54419c080f015eb7507a/packages/example-1/vite.config.ts#L5

**Comment by R-Bower** [comment]
Sorry for this thread!  I should've read through the discussions.
[This issue](https://github.com/seek-oss/vanilla-extract/discussions/281) is a much better thread.
